### DISCUSSION TITLE
How do I inject DurableTaskClient into services?

### DISCUSSION THREAD

**Discussion by Tiberriver256** [Azure/azure-functions-durable-extension#2934] [Q&A] [answered]
If I'm Azure Functions Isolated and this nuget package:

```xml
<PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.1.7" />
```

I can get access to a DurableTaskClient in a function with this binding:

```cs
[DurableClient] DurableTaskClient client
```

How do I get access to the same client in other services?

For example, I have a class like this:

```cs
public class MyService(DurableTaskClient client) {
 . . .
}
```

And I have it plugged into my services like:

```cs
services.AddScoped<MyService>()
```

I get the following error message:

```
Unable to resolve service for type 'Microsoft.DurableTask.Client.Entities.DurableEntityClient' while attempting to activate 'MyService'
```

**Comment by cgillum** [comment]
By other services, do you mean like other than Azure Functions (such as App Service, AKS, etc.)? This is possible, but not really documented.

@jviau by chance do you have a sample that shows how to create external clients using `Microsoft.DurableTask.Client` and the orchestration service client shim?

**Reply by Tiberriver256** [reply]
Actually, I meant services within Azure Functions.

For example, this is what I'd refer to as a service with a dependency on DurableTaskClient:

```cs
public class MyService(DurableTaskClient client) {
 . . .
}
```

I found out I could get dependency injection working by adding this to my startup:

```cs
services.AddDurableTaskClient(c =>
{
    c.UseGrpc();
});
```

That is working for me at the moment. Thank you!

**Reply by jviau** [reply, marked as answer]
This is not possible today in .NET isolated, as the server information for the gRPC client is only included in the binding payload itself. When Durable moves to using the functions host provided gRPC server, then it will be known at startup and the client can be included in the service container.

**Reply by Tiberriver256** [reply]
Shucks :(

It does indeed not work. Everything worked fine locally with func.exe but pushing up into Azure Functions things didn't quite work right. The strange part is I didn't see any exceptions but it never created the tables in our storage account or anything.

Thanks for the assist!

**Reply by ltouro** [reply]
If you are using an Azure Storage provider, maybe this can give you some pointers:

https://github.com/Azure/azure-functions-durable-extension/blob/dev/src/WebJobs.Extensions.DurableTask/DurableTaskJobHostConfigurationExtensions.cs#L40C34-L40C35

Maybe this is not applicable to Isolated worker. In this case, sorry for the noise. But I use that to communicate with a custom worker runtime with no problem.
